### DISCUSSION TITLE
Convert admin area help files to LAN strings?

### DISCUSSION THREAD

**Discussion by rica-carv** [e107inc/e107#5240] [Ideas]
Currently, the admin area help text ](https://github.com/e107inc/e107/tree/master/e107_languages/English/admin/help) is encoded in a diferent files inside folder help in admin area of language files.
Couldn't these strings be converted to the "normal" LAN strings inside the LAN files?
Just an ideia, to consolidate and make LAN strings and files more manageable....

**Comment by Vodhin** [comment]
I think this is a remnant from the beginnings of e107 - I think the Help LAN files are keyed to the Admin Area using matched name.php files: It simply loads `"e107_languages/".e_LANGUAGE."/help/".e_PAGE ` and gets the correct Caption and Text for the sidebar. As such, there is really no need to convert to LAN strings, and would be rather daunting since there are many languages available.

**Reply by rica-carv** [reply]
Ok, i aknowledge that, but it's quite ankward to be unable to edit those files via the admin area in E107, since they are not LAN strings.

I know it would be troublesome the conversion, but it's better to edit all languages inside e107, or to have two ways to edit those language files (ie, the LAN files in e107, and the php help files in a external editor)?

And isn't there also daunting to update all the language files when new strings are entered in the English LAN files?

Furthermore, there's already a help file relying on LAN: [Cron.php](https://github.com/e107inc/e107/blob/master/e107_languages/English/admin/help/cron.php)

And for performance, it would also be a step ahead: Instead of loading another file from the server because one more text string, the text string would already be in the LAN constants file already loaded....

I can't possibly see any downfall for simplifying the admin help strings....

**Reply by tgtje** [reply]
Ok, i can follow the idea behind it, yes could be good point. No argue there.

However : in the last 4 years only 1 change was made to the help section (7 mnths ago).
So question arises : do all the work to create new stuff to be able to use LAN's o texts that hardly change ? (description use).

In that i agree with @Vodhin tables >specially  css and or added js ; on large textual content, calls, url pointers, çommand like things etc...
And yes would be possible and easy for some files ; but _only would benefit_ if ALL in help could be recoded etc.. and do not see that happening in a short period. Will 1 remain : no solution..  
(well.. my opinion.  If it ain’t broke, don’t fix it is not the point.. some things might/could be better hence also the idea i believe.
let's find out how others might see it ?)

**Reply by rica-carv** [reply]
Ok, sorry for asking....

But beware, there's bugs on those help files when you use another languages. I had to fix the Portuguese ones the hard way because the help strings, as they are in English, can't allow for accented characters in the displayed help string....

Regular LAN strings don't have this issue.... Why?

But ok, not a critical error....

**Reply by tgtje** [reply]
**Do not be sorry, the idea has 'ground'** , but the methods to get to such point is labour extensive.
As for the chars : (used iso 8859 i saw):

Maybe the editor in use ?
However knowing (from reading) Portuguese is not an easy one.

Basically most files are in UTF 8;  best to work with UNIX (format) Unicode UTF8 no BOM (coding).
Ever given it a try?  just in case.

Using a translation engine (parts of a file with  � ) and then into fresh php file using above gives 

`A partir daqui voc� pode permitir/proibir que os usu�rios postem imagens no site, definam o m�todo de redimensionamento e visualizem avatares carregados.`  orgin. 
in tranlator switching languages back and forth (as i do not know Portuguese)  will give me this and pasting in editor : 
 `A partir daqui você pode permitir/proibir que usuários postem imagens no site, definir o método de redimensionamento e visualizar avatares carregados.` 
  and is displayed correctly . 

( From here you can allow / disallow the ability for users to post images on the site, set resize method and view uploaded avatars.)

**Reply by tgtje** [reply]
Just a quick answer ,, Maybe a possibility is there. 
Will get back to it, however _can not give a time frame_.. 1 'little' test showed promising result..
Figuring out some interaction between files (what, were, how etc..)

![testing2](https://github.com/e107inc/e107/assets/1811857/0404fb1e-5e5f-452a-ae63-8ef3dc083ce1)

testresult : not plain as from help text but by a LAN call.

**Reply by Vodhin** [reply]
I use [PSPad ](http://www.pspad.com) for my coding. It has a built in FTP client so I can edit anything on the server as live and see the Production results. It's just an enhanced Text Editor so you get to see everything going on. oh, and it is free.

**Reply by rica-carv** [reply]
I also use PSPad. So nothing weird here.

The isue is that LAN constants are correctly rendered in the original accented language, but the help files are not, so they need an additional tweaking....

So i wonder if the help files were converted to regular LAN constants, if the issue would remain....

**Reply by tgtje** [reply]
@rica-carv just guessing here : you use the original files ?
If so ( i can not tell what you precisly do on your pc, and or its running system operating language.
I mention it just because it is hard to tell what is happening.
Example : suppose pspad has other format, and coding, an you would open a help file for editting...

I am not sure if the help file is actually utf 8 no BOM and UNIX, and assume your pspad is DOS or ANSI using..
Next you paste Potuguese in it and save it.. Question remains > what type does it become...

If it becomes a mix of both, you are/have double encoding  (that could be the reason for � .
Issue on background why could be BYTE missing so 'reading expects, but something is gone so  � .

Just try : make a copy of the help file in **RAW** mode from github. (copy/pasting)
_PsPad new file_  UNIX Unicode UTF8 no BOM >paste it in there. add your translation and save it with that same format/coding.. Is there a difference ?   (without the mbstring...)

**Reply by rica-carv** [reply]
@tgtje I copyed exactly the same file from the English language, both are UNIX, so no format changed for the file.
I got the issue with the original format, so i need to use the forced enconding.

It would be weird to be checking the format of the files everytime it's edited.....

**Reply by Vodhin** [reply]
If you open the file on the server PSPad default to the encoding that the file uses. My theory of what is happening is that these files are sent directly to the parser via `$ns -> tablerender("caption",$text)` (through e_render_class.php), and - I think -  LAN strings get a bit of extra parsing from being defined as a Constant (via PHP?). 

Perhaps:

```
$text = "blah blah bla...";
$ns -> tablerender("caption",$text);
```

could be 

```
define("THE_HELP_FILENAME","blah blah bla...");
$ns -> tablerender("caption",THE_HELP_FILENAME);
```

**Reply by tgtje** [reply]
Indeed would be weird, was a question/remark. @rica-carv 
I don not know @Vodhin , i will have to test.
Anyway..quick glances : a lot of it isn't used at all it seems... have no 0.6-7 or v1 running so can't check for BC without investigating.
Some are in use, some only caption, others caption and texts..

**Reply by Jimmi08** [reply]
Just note. I was a long-time user of PSPad. After its updates it started to have issues with coding, nothing from their support forum didn't help - for my language this was really important...  I moved to Visual Code and no issues anymore.  It was nightmare at that time and breaking my code.

to e107 - there is somewhere a list of supported national characters.  @rica-carv did you check it for your language?  I remember that I had to add some for Czech or Slovak years ago to have it fully working in e107 environment.  Just reminder.

For me, there are more serious issues in e107 than this. I am slowly "retiring" for lack of a better word because after so many years basic things are not finalized.  

@rica-carv it is nice to see you again

**Reply by rica-carv** [reply]
@Vodhin Precisely my point, the LAN constants have different parsing than the help hard coded files.... So if they became LAN, maybe the issue is gonne?

@tgtje I don't get it, but if you're talking about the help files, yes, they are all used in the current admin interface....

@Jimmi08 First, thanks for the (re)welcome :wink:. I'm trying to get back and put my site a second chance... Let's see how it works... Lot of rewriting to do....
Second, i think e107 (or PHP) already supports the national characters, because the LAN constants are correclty rendered with the accent chars, so i think its not an issue from there. But thanks for the tip, anyway....

**Reply by tgtje** [reply]
@rica-carv for understanding i compare dutch files vs english etc...  > so investigating > : Dutch has 'still' all files (eg : displays more than todays core uses). > dutch have more context.. (if yes/no needed ?? thats a discussion for dutch ...).
Just to inform you... (no errors for that anyway).

A typical file that is confusing is > help/banlist vs the admin displayed help blocks.. hardcoded LAN's used in prefs etc... 
Question arises : does it make the file obsolete ? That has also to be taken in account..

That said.. when playing with LAN's i have to be sure it works without errors (multi lang)..
And if it all works.. hmm sure... discussion will arise due to nature of LAN calls, do they (dev's etc..) like a lot of texts in the core lan file(s)?  .
Amount of text data will remain same (storage size) but LAN's will increase...

**Reply by Vodhin** [reply]
@rica-carv it may not be PHP doing the parsing: 
`define("MY_CONSTANT_STRING","My Constant String")` outputs a $tring_var array equivalent to `$GLOBALS['MY_CONSTANT_STRING'] ` which PHP is passing pretty much the same way as any other $tring_var, such as the `$text="Some help text"` that the Help files are using. I believe that e107 went with that method to avoid creating long string Constants saved into PHP memory.

The only difference I can see between these two is most LAN Constants are added to text that usually get added to the page's $text string which in turn should be passed through `e107::getParser()->toHTML($text) `before sending to the `$ns->tablerender("Caption",$text)` function. The help files do not seem to be passes through the parser. 

Maybe all that is needed is pass the help page's $text through the parser? 
`$ns->tablerender("Caption",e107::getParser()->toHTML($text));`

**Reply by tgtje** [reply]
Yes @Vodhin that is/can be coded as in this example used on help/cron.  and changed into admin/cron php as : 
`		function renderHelp()
		{
			return array('caption'=>LAN_HELP, 'text'=>e107::getParser()->toHTML(LAN_CRON_64, true));
		}
		`
However.. it can be done easier (first tests shows it) _but conditional_..  
Meaning : as long as it stays inside the admin area, i am able to do it be merely LAN's no more no less.
![lans](https://github.com/e107inc/e107/assets/1811857/758e6958-8992-4e4f-a68b-0148e381ff72)

In this case (busy with it) i only need new LAN defines in the corresponding admin/lan files php
where  (image) CUSLAN_84 = text  and CUSLAN_83 is caption  (file belonging : admin/lan_cpage. (testcase)
functional...

Even could be more short, but after thinking about it (as long no hardcoding is used or future). The next will work to, but one looses the 'what does it represent.. so i prefer the first image.
![fulllan](https://github.com/e107inc/e107/assets/1811857/80917ca5-e306-4c77-98b2-cc11c51248d3)

edit: i like to keep it as simple as possible without interfering (by extra coding). 
Ok the help folder remains but the content is changed by LAN's and (text) goes to language define.
If this will work for all ..have to find out... (as mentioned earlier : obsolete parts??),
 and for future (suppose help folder gone)  : for coder it is copy paste (since defines exist) to add render in system file (if needed).

**Reply by rica-carv** [reply]
@tgtje  That's what i was talking about from the beggining, move the help files text to LAN constants on the language files....

**Reply by tgtje** [reply]
Yep @rica-carv correct, but could not say on forehand if it would be possible  (reason behind : some texts only appear after switching things, that could make it difficult). 100% of all ??  still remains to be seen. (had to think if coded vs other methods), and have to follow some rules of course.

Also have to use different language packs to see the function.. that adds to work (and hopefully no errors)
I am working on it (when time allows ) but let's hope, when finished, it will be accepted in this way,.. ;-)
b.t.w : i use git version, but should not be of influence.

**Reply by rica-carv** [reply]
> reason behind : some texts only appear after switching things, that could make it difficult

@tgtje I don't thinks is that hard, a simple code move will solve the issue.... Plus the definition of it's LAN constants....

**Reply by tgtje** [reply]
@rica-carv ..well on its way..

**Reply by tgtje** [reply]
@rica-carv  willing to test? help = utf 8 based)  [please read readme](https://github.com/tgtje/testlan/tree/main)
PT files used from latest commit.

**Reply by rica-carv** [reply]
@tgtje Thank's, i'll give a look this weekend. Thanks

**Comment by rica-carv** [comment]
@tgtje I've tested it, and for me LAN's are working...

There's just missing files to be moved:
admin_log.php, used in e107_admin/admin_log.php
banlist.php, used in e107_admin/banlist.php
image.php, used in e107_admin/image.php
mailout.php, used in e107_admin/mailout.php
menus.php, used in e107_admin/menus.php (the help on this page is show diferently from all the other, theres a ? icon that shows the help string...)
newspost.php, cat & pref are not used in  e107_admin/newspost.php, althought the subpage exists... Weird...
theme.php, used in e107_admin/theme.php
upload.php, not used?
userclass2.php, not used?
users.php, not used?

Now there's a small thing on your Portuguese files, i have downloaded them and the accent words come like �, but when shown in admin area, they are correctly accented chars....

Probably after, also the PHP code inside the old help files is redundant and can be moved to main files....

**Comment by tgtje** [comment]
@rica-carv  Thank you for testing

admin_log.php, used in e107_admin/admin_log.php; no use
banlist.php, used in e107_admin/banlist.php; no use
image.php, used in e107_admin/image.php; no use
menus.php, used in e107_admin/menus.php ; correct when ? but no lans inside file (still ANSI)
theme.php obsolete (still in PT pack)

mailout.php, used in e107_admin/mailout.php   9 lan's used 4 (texts) not  )still ANSI i saw > format should be UTF8
newspost.php 3 lan's used  2 (texts) not > UTF8
upload.php 2 lans used 2 (texts) not  > UTF8
userclass2.php, 6 lans used, 3 (texts) not > UTF8
users.php, 2 lans used  4 (texts) not > UTF8

first : should like to know in which/or from file they come > accent words come like � (only changed files to UTF8, the download codepack has them in ANSI)

And YES of course : the texts not used... > this is where it stops : 
due to the long term existance of the folder, and called for content to display, it is for 'me' (unknown) what the actual usage
or placement was created for (v0.8 development > (starting in 0.7 ;  2010 ) .
It is not up to me to make changes to core files (and likely unknown if i will succeed anyway - no real coder- struggling with renderHelp > function/private/public and position, tests keep failing)

**In short : LAN's/defines are possible, but original files remain in place (with lan use). and 'calls are unchanged.)**

So this could be step 1 ?
And a coder uses the LAN's (as they would be present) to render them in core file, and the  help files become obsolete..

What would be best way to go from here?.

**Reply by rica-carv** [reply]
@tgtje There's something wrong there, or my e107 is outdated, or something else, because in my e107 v2.3.3:
admin_log.php, is used in e107_admin/admin_log.php, right below the left side admin menu
banlist.php, used in e107_admin/banlist.php, again right below left side admin menu
image.php, used in e107_admin/image.php below left side admin menu
mailout.php, used in e107_admin/mailout.php below left side admin menu
menus.php, used in e107_admin/menus.php (the help on this page is show diferently from all the other, theres a ? icon that shows the help string...)
theme.php, used in e107_admin/theme.php below left side admin menu

Further, if LAN defines are possible, the PHP code files are redundant, because they will be always the same code in any language, and we are duplicating, triplicating code among languages. Right?
So, if all the help strings are gone to LAN's, what's the point on having more files without any language string, only code, to show them. For me, the next step is move the code to inside the core files. For me it's the logical step to go now....

**Reply by tgtje** [reply]
@rica-carv 
oops well being honest :
admin_log missed only active when log is enabled 

rest :  plain overseen probably worked too fast

mailout ? 4 lines i see not in use ?? rest lan
help/theme php is no longer in system (_gone = Ajuda do Gerenciador de Temas_; what remains is the block for : Filtro de visibilidade)
_have to work up a little to have them complete_.

Next step would indeed be code (eg rendering the help blocks)
will still try..
